src/main/java/org/team1540/base/util/AsyncCommand.java

```diff
@@ -0,0 +1,131 @@
+package org.team1540.base.util;
+
+import edu.wpi.first.wpilibj.Notifier;
+import edu.wpi.first.wpilibj.command.Command;
+import org.jetbrains.annotations.Contract;
+
+/**
+ * A {@link Command} that spawns off a new {@link Notifier} on initialization and stops it. That
+ * thread will call the {@link #runPeriodic()} ()} method at a user-specified interval with precise
+ * timings.
+ *
+ * @see Thread
+ */
+public abstract class AsyncCommand extends Command {
+
+  private long interval;
+  private boolean finished = false;
+
+  private Notifier notifier;
+
+  /**
+   * Constructs a new {@code AsyncCommand} with a preset periodic interval.
+   *
+   * @param interval The interval between {@link #run()} calls, in milliseconds.
+   */
+  public AsyncCommand(long interval) {
+    this.interval = interval;
+  }
+
+  /**
+   * Runs the action of this command.
+   * <p>
+   * This is equivalent to the {@link #execute()} method in normal commands. Note that this method
+   * will be called in a seperate thread, and as such any public methods it calls, fields it uses,
+   * etc. MUST be thread-safe.
+   */
+  protected void runPeriodic() {
+
+  }
+
+  /**
+   * Runs initialization code for this command.
+   *
+   * This is equivalent to the {@link Command#initialize()} method.
+   */
+  protected void runInitial() {
+
+  }
+
+  /**
+   * Called when the {@code AsyncCommand} ends peacefully. This is equivalent to {@link
+   * Command#end()}.
+   *
+   * This method is called when the command ends peacefully, i.e. when {@link #markAsFinished()} was
+   * called. However, it is possible that another command interrupts this command after {@link
+   * #markAsFinished()} was called but before the command was removed from the scheduler, and as
+   * such this method might not be called even if {@link #markAsFinished()} was called.
+   */
+  protected void runEnd() {
+
+  }
+
+  /**
+   * Called when the {@code AsyncCommand} is interrupted.
+   *
+   * This method serves an identical purpose to {@link Command#interrupted()}. By default, this
+   * method calls {@link #runEnd()}.
+   */
+  protected void runInterrupt() {
+    runEnd();
+  }
+
+  @Override
+  protected final void initialize() {
+    finished = false;
+    runInitial();
+
+    notifier = new Notifier(() -> {
```

Comment:
Didn't know you could make a Runnable like this, cool

Reply:
https://www.geeksforgeeks.org/functional-interfaces-java/